File: apps/client/src/lib/schemas/mcp.ts
import { z } from "zod";

/** Runtime validator for `mcp_tool_docs` (`screens/mcp.tsx`) — the MCP
 * contract's JSON Schema tool list (`tt_mcp::tool_definitions`), worth
 * catching a shape drift on since it's rendered directly as documentation. */

const McpToolParamSchema = z.object({
  type: z.string().optional(),
  description: z.string().optional(),
  enum: z.array(z.string()).optional(),
});

export const McpToolDocSchema = z.object({
  name: z.string(),
  description: z.string(),
  inputSchema: z.object({
    type: z.string(),
    properties: z.record(z.string(), McpToolParamSchema).default({}),
    required: z.array(z.string()).default([]),
  }),
  /** MCP's own tool annotations. The server emits `readOnlyHint: false` on the
   * tools that write and omits the whole block otherwise, so absence means
   * "no claim made", not "read-only". */
  annotations: z.object({ readOnlyHint: z.boolean().optional() }).optional(),
});

export const McpToolDocsSchema = z.array(McpToolDocSchema);

export type McpToolDoc = z.infer<typeof McpToolDocSchema>;

/** Runtime validator for `mcp_status` — whether *this* instance won the bind
 * race for the MCP port, and which port that is. */
export const McpStatusSchema = z.object({ serving: z.boolean(), port: z.number() });

export type McpStatus = z.infer<typeof McpStatusSchema>;

/** Runtime validator for `mcp_test_call` — what one real round-trip against the
 * MCP endpoint came back with. A refusal is a result to display, not an error:
 * the point is seeing what a client would see. `sentOrigin` records whether the
 * request deliberately carried an `Origin` header. */
export const McpTestResultSchema = z.object({
  status: z.number(),
  body: z.string(),
  durationMs: z.number(),
  sentOrigin: z.boolean(),
});

export type McpTestResult = z.infer<typeof McpTestResultSchema>;

File: crates-tauri/tt-app/src/mcp.rs
//! Tauri bridge for the MCP screen's tool documentation. The server runs in
//! this process ([`crate::mcp_http`]), so the screen reads the tool list
//! straight from [`tt_mcp::tool_definitions`] — the same source `tools/list`
//! answers from — instead of round-tripping a request through the socket. It
//! is also the only way the screen can show the contract when this instance
//! lost the bind race and is serving nothing.

use serde_json::Value;

/// The full JSON Schema tool list (name, description, inputSchema) from the
/// MCP contract — identical to what `tools/list` returns to a client.
#[tauri::command]
pub fn mcp_tool_docs() -> Value {
    tt_mcp::tool_definitions()
}
